### DISCUSSION TITLE
Approach to struct types and docs

### DISCUSSION THREAD

**Discussion by zakhenry** [flux-rs/flux#452] [General]
I just watched the talk linked in reddit (https://www.reddit.com/r/rust/comments/13l6s4l/liquid_types_for_rust/) and found the discussion about the meaningfulness of the declared parameters for the struct `refined_by` macro to be quite valid as it was a little confusing. I tried to make sense of it from the docs but alas there no page on structs just yet. Had a stab at getting it working in the playground with a somewhat real world solution and I think I got there in the end (at least I'm getting errors when I expect them). Is there a cleaner way that I could achieve this same outcome?

```rust

#[flux::refined_by(width: int, height: int)]
struct Canvas {
    #[flux::field({i32[@width] | width > 0})]
    width: i32,
    #[flux::field({i32[@height] | height > 0})]
    height: i32,
}

impl Canvas {

    #[flux::sig(fn(width: {i32[@w] | w > 0}, height: {i32[@h] | h > 0}) -> Canvas[w, h])]
    pub fn new(width: i32, height: i32) -> Canvas {
        Canvas{ width, height }
    }

    #[flux::sig(fn(&mut Canvas[@w, @h], {i32[@x] | x >= 0 && x <= w - 1}, {i32[@y] | y >= 0 && y <= h - 1}, bool))]
    fn set_pixel(&mut self, x: i32, y: i32, on: bool) {
        // todo implement
    }

}

fn main() {
   let mut canvas_too_small = Canvas::new(0, 10); // fail
   let mut canvas = Canvas::new(8, 4);

   canvas.set_pixel(2, 12, true); // fail
   canvas.set_pixel(8, 4, true); // fail
   canvas.set_pixel(2, 3, true); // pass
}
```

I feel like the refinement signature on `set_pixel` can probably be simplified?

Also feel free to use this example in the docs if it helps any, I was trying to come up with an example that should be somewhat familiar to most developers and would feel more "real world". Hoping to show this awesome initiative off to colleagues and coming up with some examples to get them excited 😁

**Comment by ranjitjhala** [comment]
Hi @zakhenry -- thanks for your interest and this fun example! Yes, we have a bunch of other examples lined up but had to hold them since the paper was under review. Your example can be simplified a bit as instead of 

`{i32[@x] | cond}` you can write `x:i32{cond}` so you could write



 like so:

```rust
    #[flux::sig(fn(width:i32{width > 0}, height:i32{height > 0}) -> Canvas[width, height])]
    pub fn new(width: i32, height: i32) -> Canvas {
        Canvas{ width, height }
    }

    #[flux::sig(fn(&mut Canvas[@w, @h], x:i32{x >= 0 && x <= w - 1}, y:i32{y >= 0 && y <= h - 1}, bool))]
    fn set_pixel(&mut self, x: i32, y: i32, on: bool) {
        // todo implement
    }
```
